File: day13/day13.cpp
#include <algorithm>
#include <fstream>
#include <vector>
#include <fmt/format.h>

using namespace std;

struct Map
{
	vector<pair<int, int>> data;

	void fold_up(int pos)
	{
		for (auto& p: data)
		{
			if (p.second > pos)
			{
				p.second = 2 * pos - p.second;
			}
		}
	}

	void fold_left(int pos)
	{
		for (auto& p: data)
		{
			if (p.first > pos)
			{
				p.first = 2 * pos - p.first;
			}
		}
	}

	size_t count(void)
	{
		sort();
		size_t count = 0;
		int px = -1;
		int py = -1;
		for (auto [x, y] : data)
		{
			if (x != px || y != py)
			{
				count++;
			}
			px = x;
			py = y;
		}
		return count;
	}

	void print(void)
	{
		sort();

		int cx = 0;
		int cy = 0;
		for (auto [x, y]: data)
		{
			if (cy == y && cx > x)
			{
				continue;
			}

			for (; cy < y; cy++)
			{
				putchar('\n');
				cx = 0;
			}

			for (; cx < x; cx++)
			{
				putchar(' ');
			}

			putchar('#');
			cx++;
		}
		putchar('\n');
	}

	void sort(void)
	{
		std::sort(data.begin(), data.end(), [](const auto& a, const auto &b) {
			return (a.second != b.second)
				? (a.second < b.second)
				: (a.first < b.first);
		});
	}
};

istream& operator>>(istream& in, Map &m)
{
	m.data.clear();

	string line;
	while (getline(in, line) && !line.empty())
	{
		auto p = line.find(',');
		if (p != string::npos)
		{
			int x = stoi(line.substr(0, p));
			int y = stoi(line.substr(p+1));
			m.data.push_back(make_pair(x, y));
		}
	}
	return in;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream in(argv[1]);
	if (!in)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	Map m;
	in >> m;

	size_t part1 = SIZE_MAX;
	string cmd;
	while (getline(in, cmd))
	{
		auto sep = cmd.find("=");
		if (sep != string::npos)
		{
			int pos = stoi(cmd.substr(sep+1));

			switch(cmd.at(sep - 1))
			{
			case 'x': m.fold_left(pos); break;
			case 'y': m.fold_up(pos); break;
			}
		}
		if (part1 == SIZE_MAX)
		{
			part1 = m.count();
		}
	}
	in.close();

	fmt::print("Part1: {}\n", part1);
	fmt::print("Part2:\n");
	m.print();
	return 0;
}

File: day13/day13.py
#!/usr/bin/env python3

class Sheet(object):
    def __init__(self, values):
        self.values = values
        self.width = max(x for x, y in values) + 1
        self.height = max(y for x, y in values) + 1

    def fold_up(self, fy):
        self.height = fy
        lst = []
        for x, y in self.values:
            if y > fy:
                lst.append((x, 2*fy-y))
            else:
                lst.append((x, y))
        self.values = lst

    def fold_left(self, fx):
        self.width = fx
        lst = []
        for x, y in self.values:
            if x > fx:
                lst.append((2*fx-x, y))
            else:
                lst.append((x, y))
        self.values = lst

    def count(self):
        return len(set(self.values))

    def __str__(self):
        s = set(self.values)
        return "\n".join(
            "".join((x, y) in s and "#" or " " for x in range(self.width))
            for y in range(self.height)
        )

def parse_sheet(txt):
    coords = []
    instr = []
    state = 0
    for line in txt.splitlines():
        if state == 0:
            if line == "":
                state = 1
            else:
                coords.append(tuple(int(x) for x in line.split(",")))
        else:
            a, b = line.split("=")
            instr.append((a[-1], int(b)))

    return Sheet(coords), instr

def solve(txt):
    s, instr = parse_sheet(txt)
    part1 = None
    for coord, pos in instr:
        if coord == "y":
            s.fold_up(pos)
        elif coord == "x":
            s.fold_left(pos)

        if part1 is None:
            part1 = s.count()

    return part1, s

import sys

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()

        part1, part2 = solve(txt)
        print("Part1:", part1)
        print("Part2:")
        print(part2)
    except FileNotFoundError:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)
